apps/ewallet/lib/ewallet/gates/activity_log_gate.ex

```diff
@@ -0,0 +1,138 @@
+defmodule EWallet.ActivityLogGate do
+  @moduledoc """
+  Handles the logic for manipulating activity logs.
+  """
+  import Ecto.Query
+  alias EWalletDB.Repo
+  alias EWallet.Web.Preloader
+  alias ActivityLogger.ActivityLog
+
+  @doc """
+  Adds the originator and target struct for each activity log in the list.
+  """
+  def add_originator_and_target(activity_logs, module_mapper) when is_list(activity_logs) do
+    activity_logs
+    |> format_preload_values(module_mapper)
+    |> query_preload(module_mapper)
+    |> add_to_original(activity_logs)
+  end
+
+  # Extract and prepare data that need preloading in the following format
+  # %{
+  #   Entity1: [uuid1, uuid2, uuid3],
+  #   Entity2: [uuid1, uuid2],
+  #   ....
+  # }
+  defp format_preload_values(data, module_mapper) do
+    Enum.reduce(data, %{}, fn activity_log, acc ->
```

Comment:
I think Elixir's immutability might make this code quite intensive to build from top down. So it might be better to use more native Enum functions + avoid nested loops altogether.

Maybe something like this instead (untested)?

```ex
defp format_preload_values(data, module_mapper) do
  {data, []}
  |> extract_entity_uuids(:originator_type, :originator_uuid, module_mapper)
  |> extract_entity_uuids(:target_type, :target_uuid, module_mapper)
  |> elem(1)
  |> Enum.uniq()
  |> Enum.group_by(&elem(&1, 0), &elem(&1, 0))
end

defp extract_entity_uuids({raw, extracted}, schema_type, record_identifier, module_mapper) do
  extracted = 
    Enum.reduce(raw, extracted, fn activity_log, acc ->
      schema = activity_log |> Map.get(schema_type) |> ActivityLog.get_schema()

      if preloadable?(schema, module_mapper) do
        [{schema, Map.get(activity_log, record_identifier} | acc]
      else
        acc
      end
    end

  {raw, extracted}
end

#
# Delete code below
#
#
# defp format_preload_values(data, module_mapper) do
#   Enum.reduce(data, %{}, fn activity_log, acc ->
#     acc
#     |> add_preload_if_valid(
#       ActivityLog.get_schema(activity_log.originator_type),
#       activity_log.originator_uuid,
#       module_mapper
#     )
#     |> add_preload_if_valid(
#       ActivityLog.get_schema(activity_log.target_type),
#       activity_log.target_uuid,
#       module_mapper
#     )
#   end)
# end
#
# # Checks if the given schema needs to be loaded.
# # If no -> do nothing
# # If yes -> adds an array of uuids for the given schema key
# defp add_preload_if_valid(preloads, schema, uuid, module_mapper) do
#  case preloadable?(schema, module_mapper) do
#    true ->
#      preloads
#      |> Map.put_new(schema, [])
#      |> add_uuid_if_not_present(schema, uuid)

#     false ->
#      preloads
#  end
# end

# # Adds the given uuid to the list for the schema if not already present
# defp add_uuid_if_not_present(preloads, schema, uuid) do
#  uuids =
#    case !Enum.member?(preloads[schema], uuid) do
#      true -> [uuid | preloads[schema]]
#      false -> preloads[schema]
#    end

#   Map.put(preloads, schema, uuids)
# end
```